File: src/core/diagram/code/constraints.rb
require 'enso'

module Constraints

	class ConstraintSystem
	  def initialize
	    @vars = {}
	    @number = 0
	  end
	  
	  def variable(name = "v#{@number}", value = nil)
	    @number += 1
	    #puts "#{name} = #{value}"
	    Variable.new(name, value)
	  end
	  
	  def value(n)
	    variable("(#{n})", n)
	  end
	end
	
	class Constant
	  def initialize(val)
	    @value = val
	  end
	  
	  
	  def add_listener(l)
	  end
	  
	  def internal_evaluate(path)
	  	@value
	  end

	  def value
	  	@value
	  end
	
		def to_i
	    value.to_i
	  end
	  
	  def to_s
	    value.nil? ? "nil" : value.to_s
	  end
	  
	  def to_ary
	    [self]
	  end
	end
	
	class Variable < Constant
	  def initialize(name, val = nil)
	    super(val)
	    @name = name
	    @dependencies = []
	    @vars = []
	    @bounds = nil
	  end
		 
	  def add(other)
	    define_result(:add, other)
	  end
		      
	  def sub(other)
	    define_result(:sub, other)
	  end
		      
	  def mul(other)
	    define_result(:mul, other)
	  end
		      
	  def div(other)
	    define_result(:div, other)
	  end

	  def round
	    define_result(:round)
	  end
		      
	  def to_int
	    define_result(:to_int)
	  end
		      
	  # special case to implement MAX behavior
	  def max(other = raise("MAX WITH UNDEFINED"))
	    #puts "#{self} MAX #{other}"
	    other.add_listener(self) if other.is_a?(Variable)
	    @bounds = [] if @bounds.nil?
	    @bounds << other
	  end
	
	  def test(a, b)
	    variable = Variable.new("test#{self.to_s}")
	    variable.internal_define(self, a, b) do |v, ra, rb|
	      #puts "EVALUATING TEST VARIABLE!!! #{a} else #{b}"
	      v.test(ra, rb)
	    end
	    variable
	  end

	  def new_var_method(&block)
	    variable = Variable.new("p#{self.to_s}")
	    variable.internal_define(self, &block)
	    variable
	  end
	
	  def define_result(m, *args)
	    raise "undefined method #{m}" unless [:add, :sub, :mul, :div, :round, :to_int].include?(m) 
	    variable = Variable.new("p#{self.to_s}#{args.to_s}")
	    #puts "#{variable}=#{self.to_s}+#{args}"
	    variable.internal_define(self, *args) do |*values|
	      do_op(m, *values)
	    end
	    variable
	  end
	  
		def do_op(op, *values)
		  case op
		  when :add   
		  	values[0] + values[1]
		  when :sub
		  	values[0] - values[1]
		  when :mul
		  	values[0] * values[1]
		  when :div
		  	values[0] / values[1]
		  when :round
		  	values[0].round
		  when :to_int
		  	values[0].to_int
		  end
		end
	  
	  def internal_define(*vars, &block)
	    raise "DOUBLE defined Var" if @block
	    @vars = vars.map do |v|
	      if v.nil?
			    raise "Unbound variable #{v.toString}"
	      elsif v.is_a?(Constant) || v.is_a?(Variable)
	        v
	      else
	        Constant.new(v)
	      end
	    end
	    @vars.each do |v|
	      v.add_listener(self)
	    end
	    #puts "#{vars}"
	    @block = block
	  end
	  
	  def add_listener(x)
	    #puts "#{x.to_s} listening to #{self.to_s}"
	    @dependencies << x
	  end
	
	  def value
	    internal_evaluate unless @value
	    @value
	  end
	  
	  def value=(x)
	    @block = nil
	    internal_notify_change
	    @value = x
	  end

		def redo_max 
		  @value = nil
		  internal_notify_change
		end

	  def internal_notify_change
	    @dependencies.each do |variable|
	      variable.internal_notify_change
	    end
	    #@block.nil means this is a hardcoded var, probably because someone assigned to it
	    @value = nil unless @block.nil?
	  end
	    	  
	  def internal_evaluate(path = [])
	    raise "circular constraint #{path.map(&:to_s)}" if path.include?(self)
	    @value = nil if @bounds
	    if @block
	      path << self
	      vals = @vars.map do |variable|
	        val = variable.internal_evaluate(path)
	        if val.nil?
	          puts "WARNING: undefined variable '#{variable}'"
	          val = 10
	        end
	        val
	      end
	      path.pop
	      @value = @block.call(*vals)
	    end
	    if @bounds
		    @bounds.each do |b|
		      if b.is_a?(Variable) 
		        val = b.value
		      else
		        val = b
		      end
		      @value = val if @value.nil? || @value < val
		    end
		  end
	    #puts "EVAL #{@name}=#{@value}"
	    @value
	  end
	end
end

File: src/js/core/diagram/code/constraints.js
'use strict'

//// Constraints ////

var cwd = process.cwd() + '/';
var Enso = require(cwd + "enso.js");

var Constraints;

class ConstraintSystem {
  static new(...args) { return new ConstraintSystem(...args) };

  constructor() {
    var self = this;
    self.vars$ = Enso.EMap.new();
    self.number$ = 0;
  };

  variable(name = Enso.S("v", self.number$), value = null) {
    var self = this;
    self.number$ = self.number$ + 1;
    return Variable.new(name, value);
  };

  value(n) {
    var self = this;
    return self.variable(Enso.S("(", n, ")"), n);
  };
};

class Constant {
  static new(...args) { return new Constant(...args) };

  constructor(val) {
    var self = this;
    self.value$ = val;
  };

  add_listener(l) {
    var self = this;
  };

  internal_evaluate(path) {
    var self = this;
    return self.value$;
  };

  value() {
    var self = this;
    return self.value$;
  };

  to_i() {
    var self = this;
    return self.value().to_i();
  };

  to_s() {
    var self = this;
    if (self.value() == null) {
      return "nil";
    } else {
      return self.value().to_s();
    }
  };

  to_ary() {
    var self = this;
    return [self];
  };
};

class Variable extends Constant {
  static new(...args) { return new Variable(...args) };

  constructor(name, val = null) {
    super(val);
    var self = this;
    self.name$ = name;
    self.dependencies$ = [];
    self.vars$ = [];
    self.bounds$ = null;
  };

  add(other) {
    var self = this;
    return self.define_result("add", other);
  };

  sub(other) {
    var self = this;
    return self.define_result("sub", other);
  };

  mul(other) {
    var self = this;
    return self.define_result("mul", other);
  };

  div(other) {
    var self = this;
    return self.define_result("div", other);
  };

  round() {
    var self = this;
    return self.define_result("round");
  };

  to_int() {
    var self = this;
    return self.define_result("to_int");
  };

  max(other = self.raise("MAX WITH UNDEFINED")) {
    var self = this;
    if (Enso.System.test_type(other, Variable)) {
      other.add_listener(self);
    }
    if (self.bounds$ == null) {
      self.bounds$ = [];
    }
    return self.bounds$.push(other);
  };

  test(a, b) {
    var self = this, variable;
    variable = Variable.new(Enso.S("test", self.to_s()));
    variable.internal_define(function(v, ra, rb) {
      return v.test(ra, rb);
    }, self, a, b);
    return variable;
  };

  new_var_method(block) {
    var self = this, variable;
    variable = Variable.new(Enso.S("p", self.to_s()));
    variable.internal_define(block, self);
    return variable;
  };

  define_result(m, ...args) {
    var self = this, variable;
    if (! ["add", "sub", "mul", "div", "round", "to_int"].include_P(m)) {
      self.raise(Enso.S("undefined method ", m));
    }
    variable = Variable.new(Enso.S("p", self.to_s(), args.to_s()));
    variable.internal_define(function(...values) {
      return self.do_op(m, ...values);
    }, self, ...args);
    return variable;
  };

  do_op(op, ...values) {
    var self = this;
    switch (op) {
      case "add":
        return values.get$(0) + values.get$(1);
      case "sub":
        return values.get$(0) - values.get$(1);
      case "mul":
        return values.get$(0) * values.get$(1);
      case "div":
        return values.get$(0) / values.get$(1);
      case "round":
        return values.get$(0).round();
      case "to_int":
        return values.get$(0).to_int();
    }
  };

  internal_define(block, ...vars) {
    var self = this;
    if (self.block$) {
      self.raise("DOUBLE defined Var");
    }
    self.vars$ = vars.map(function(v) {
      if (v == null) {
        return self.raise(Enso.S("Unbound variable ", v.toString()));
      } else if (Enso.System.test_type(v, Constant) || Enso.System.test_type(v, Variable)) {
        return v;
      } else {
        return Constant.new(v);
      }
    });
    self.vars$.each(function(v) {
      return v.add_listener(self);
    });
    return self.block$ = block;
  };

  add_listener(x) {
    var self = this;
    return self.dependencies$.push(x);
  };

  value() {
    var self = this;
    if (! self.value$) {
      self.internal_evaluate();
    }
    return self.value$;
  };

  set_value(x) {
    var self = this;
    self.block$ = null;
    self.internal_notify_change();
    return self.value$ = x;
  };

  redo_max() {
    var self = this;
    self.value$ = null;
    return self.internal_notify_change();
  };

  internal_notify_change() {
    var self = this;
    self.dependencies$.each(function(variable) {
      return variable.internal_notify_change();
    });
    if (! (self.block$ == null)) {
      return self.value$ = null;
    }
  };

  internal_evaluate(path = []) {
    var self = this, val, vals;
    if (path.include_P(self)) {
      self.raise(Enso.S("circular constraint ", path.map(function(x) {
        return x.to_s();
      })));
    }
    if (self.bounds$) {
      self.value$ = null;
    }
    if (self.block$) {
      path.push(self);
      vals = self.vars$.map(function(variable) {
        val = variable.internal_evaluate(path);
        if (val == null) {
          puts(Enso.S("WARNING: undefined variable '", variable, "'"));
          val = 10;
        }
        return val;
      });
      path.pop();
      self.value$ = self.block$(...vals);
    }
    if (self.bounds$) {
      self.bounds$.each(function(b) {
        if (Enso.System.test_type(b, Variable)) {
          val = b.value();
        } else {
          val = b;
        }
        if (self.value$ == null || self.value$ < val) {
          return self.value$ = val;
        }
      });
    }
    return self.value$;
  };
};

Constraints = {
  ConstraintSystem: ConstraintSystem,
  Constant: Constant,
  Variable: Variable,
};
module.exports = Constraints ;
